<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Shaun Wilkinson" />

<meta name="date" content="2018-04-13" />

<title>phylogram: dendrograms for evolutionary analysis</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">phylogram: dendrograms for evolutionary analysis</h1>
<h4 class="author"><em>Shaun Wilkinson</em></h4>
<h4 class="date"><em>2018-04-13</em></h4>



<hr />
<p><strong>Important Note:</strong> From version 2.0 all <em>k</em>-mer counting and clustering functions have been migrated to the <strong>kmer</strong> package. Apologies for any inconvenience.</p>
<div id="abstract" class="section level2">
<h2>Abstract</h2>
<p>The <strong>phylogram</strong> R package is a Newick parser and tool for developing evolutionary trees as deeply-nested lists known as “dendrogram” objects. It provides functions for importing and exporting trees in parenthetic text format, as well as several tools for command-line tree manipulation. The <strong>phylogram</strong> package is released under the GPL-3 license, and is available for download from CRAN <a href="https://CRAN.R-project.org/package=phylogram" class="uri">https://CRAN.R-project.org/package=phylogram</a> and github <a href="https://github.com/shaunpwilkinson/phylogram" class="uri">https://github.com/shaunpwilkinson/phylogram</a>.<br />
<strong>Contact:</strong> shaunpwilkinson@gmail.com</p>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The R environment continues to gain popularity as a platform for working with evolutionary trees, due to the reproducible code-based workflow and the many powerful analytical tools available in a suite of open-source packages such as <strong>ape</strong> <span class="citation">(Paradis <em>et al.</em>, 2004)</span>, <strong>phangorn</strong> <span class="citation">(Schliep, 2011)</span> and <strong>Phytools</strong> <span class="citation">(Revell, 2012)</span>. These packages typically employ a tree structure known as the “phylo” object, whose primary element is an integer matrix with one row for each edge in the graph, and two columns giving the indices of the connecting nodes.</p>
<p>An alternative tree structure is the “dendrogram” object, generated using the <code>as.dendrogram</code> function in the <strong>stats</strong> package <span class="citation">(R Core Team, 2017)</span>. Rather than a matrix of edges, a dendrogram is a hierarchical list. These ‘lists of lists’ can be deeply nested, with the limit depending on the C stack size (settable <em>via</em> <code>options(&quot;expressions&quot;)</code>). A useful feature of this representation is its modularity, whereby the sub tree of a tree is itself a tree - a dendrogram within a dendrogram. This means that dendrogram objects are subsettable in the same way that standard lists are, which in addition to the inbuilt editing functions such as <code>cut</code> and <code>merge</code>, facilitates intuitive command-line tree manipulation. An especially powerful feature of this object type is that tree editing operations can be carried out recursively using fast inbuilt functions in the “apply” family such as <code>dendrapply</code> and <code>lapply</code>.</p>
<p>Each node of a dendrogram object has the following mandatory attributes:</p>
<ul>
<li>“height” the position of the node along the vertical axis (assuming the graph is orientated vertically)</li>
<li>“midpoint” the horizontal distance of the node from the left-most member of the sub tree (where the horizontal distance between adjacent leaves is 1 unit)</li>
<li>“members” the number of terminal leaf nodes belonging to the node</li>
<li>“class” all nodes have the class attribute “dendrogram”</li>
</ul>
<p>Rather than lists, terminal leaf nodes are length-1 integer vectors whose values correspond to the indices of the members in the set. The “members” attributes of leaf nodes is always 1, the “midpoint” attribute is 0, and they have two additional attributes:</p>
<ul>
<li>“leaf” TRUE for terminal nodes (NULL otherwise)</li>
<li>“label” an optional character string giving the name of the taxon or group</li>
</ul>
<p>Aside from those listed above, users may attach other objects as attributes to the dendrogram nodes. For example, “label” attributes can be attached to inner nodes, and users can specify plotting parameters for each node by setting the attributes “nodePar” and “edgePar”.</p>
<div id="example-1-build-a-dendrogram-object-manually" class="section level4">
<h4>Example 1: Build a dendrogram object manually</h4>
<p>Consider the simple example of a tree with three members named “A”, “B” and “C”, where “B” and “C” are more closely related to each other than they are to “A”. An unweighted Newick string for this tree would be <em>(A,(B,C));</em> We can manually create a dendrogram object for this basic phylogeny and plot the tree as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dv">1</span>, <span class="kw">list</span>(<span class="dv">2</span>, <span class="dv">3</span>))
## attach &quot;leaf&quot; and &quot;label&quot; attributes to leaf nodes
<span class="kw">attr</span>(x[[<span class="dv">1</span>]], <span class="st">&quot;leaf&quot;</span>) &lt;-<span class="st"> </span><span class="ot">TRUE</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">1</span>]], <span class="st">&quot;leaf&quot;</span>) &lt;-<span class="st"> </span><span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">2</span>]], <span class="st">&quot;leaf&quot;</span>) &lt;-<span class="st"> </span><span class="ot">TRUE</span>
<span class="kw">attr</span>(x[[<span class="dv">1</span>]], <span class="st">&quot;label&quot;</span>) &lt;-<span class="st"> &quot;A&quot;</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">1</span>]], <span class="st">&quot;label&quot;</span>) &lt;-<span class="st"> &quot;B&quot;</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">2</span>]], <span class="st">&quot;label&quot;</span>) &lt;-<span class="st"> &quot;C&quot;</span>
## set &quot;height&quot; attributes for all nodes
<span class="kw">attr</span>(x, <span class="st">&quot;height&quot;</span>) &lt;-<span class="st"> </span><span class="dv">1</span>
<span class="kw">attr</span>(x[[<span class="dv">1</span>]], <span class="st">&quot;height&quot;</span>) &lt;-<span class="st"> </span><span class="dv">0</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]], <span class="st">&quot;height&quot;</span>) &lt;-<span class="st"> </span><span class="fl">0.5</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">1</span>]], <span class="st">&quot;height&quot;</span>) &lt;-<span class="st"> </span><span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">2</span>]], <span class="st">&quot;height&quot;</span>) &lt;-<span class="st"> </span><span class="dv">0</span>
## set &quot;midpoints&quot; attributes for all nodes
<span class="kw">attr</span>(x, <span class="st">&quot;midpoint&quot;</span>) &lt;-<span class="st"> </span><span class="fl">0.75</span>
<span class="kw">attr</span>(x[[<span class="dv">1</span>]], <span class="st">&quot;midpoint&quot;</span>) &lt;-<span class="st"> </span><span class="dv">0</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]], <span class="st">&quot;midpoint&quot;</span>) &lt;-<span class="st"> </span><span class="fl">0.5</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">1</span>]], <span class="st">&quot;midpoint&quot;</span>) &lt;-<span class="st"> </span><span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">2</span>]], <span class="st">&quot;midpoint&quot;</span>) &lt;-<span class="st"> </span><span class="dv">0</span>
## set &quot;members&quot; attributes for all nodes
<span class="kw">attr</span>(x, <span class="st">&quot;members&quot;</span>) &lt;-<span class="st"> </span><span class="dv">3</span>
<span class="kw">attr</span>(x[[<span class="dv">1</span>]], <span class="st">&quot;members&quot;</span>) &lt;-<span class="st"> </span><span class="dv">1</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]], <span class="st">&quot;members&quot;</span>) &lt;-<span class="st"> </span><span class="dv">2</span>
<span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">1</span>]], <span class="st">&quot;members&quot;</span>) &lt;-<span class="st"> </span><span class="kw">attr</span>(x[[<span class="dv">2</span>]][[<span class="dv">2</span>]], <span class="st">&quot;members&quot;</span>) &lt;-<span class="st"> </span><span class="dv">1</span>
## set class as &quot;dendrogram&quot; 
## Note that setting the class for the root node
## automatically sets the class of all nested subnodes
<span class="kw">class</span>(x) &lt;-<span class="st"> &quot;dendrogram&quot;</span>

### plot the dendrogram
<span class="kw">plot</span>(x, <span class="dt">yaxt =</span> <span class="st">&quot;n&quot;</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABm1BMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgKCgoLCwsMDAwODg4QEBARERESEhIWFhYaGhobGxsdHR0eHh4gICAhISEiIiInJycqKiorKyssLCwtLS0uLi4vLy8xMTE0NDQ2NjY3Nzc5OTk6Ojo9PT0/Pz9CQkJDQ0NGRkZISEhLS0tMTExNTU1OTk5QUFBSUlJVVVVYWFhaWlpgYGBlZWVmZmZnZ2dra2tvb29wcHBxcXFzc3N0dHR3d3d5eXl6enp/f3+BgYGCgoKDg4OEhISGhoaIiIiLi4uMjIyNjY2Ojo6Pj4+RkZGWlpaXl5eYmJiZmZmcnJydnZ2enp6fn5+goKChoaGlpaWmpqanp6eoqKipqamqqqqrq6uurq6wsLCxsbGysrK1tbW2tra3t7e7u7u9vb2/v7/Dw8PFxcXHx8fJycnMzMzNzc3Ozs7R0dHS0tLY2NjZ2dna2trd3d3k5OTm5ubo6Ojp6enq6urr6+vx8fHy8vLz8/P09PT19fX29vb4+Pj5+fn6+vr8/Pz9/f3+/v7///99kyjFAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAEL0lEQVR4nO3dV1MUQRRA4TYrKooRs6KsEXPAnMWcwYg5YQ4gKua087NdLakysJyerWEQ63wP7FTtdPet87CBBwiJehX6e4B/nYGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgUDOgY7Xr4qxpC7qtoZVfT9xzoF2NrTGaDkYddueWX0/cc6BDmzJcre7C7LcrWcGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIGAgYCBgIDIVB7RxmX5pR7puNlVhMPgEDHq6aWUTOs3DOTxmY18QAIdK0h/ZqntenX9MxAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAIOdABw3Uu/ZH6dd03U6/5uOV9Gt65n/qBQYCBgIGAgYCBgIGAgYC/RHo881+OLRSuQf6cnvduPhDO1uuv08uzqiuay3GH/HsROlHS+Ot9MP1IN9AX9sax4cRy07H3v+wOoTZNwYt3DQ/nI0+Y0eYUnporgprv1Y05O9yDFS8v3lCGBWa3sQvWTzx0fPlQzaU1i6dHbvmTNja9f3x465wMv2Qf8kv0LbJoWr15VehLcWa6t1J8iBcK10dHRq7Zu6y7qsVdWnmKyO/QGH0sfdJ8jZVoHAqSV6E7y8mzdGDjjnSfXUkiz9DlV+ghYOGLm/9kDJQc5J0pQ20v/tqX02Ko8qOkMEekToPzwtj1oQ7KZZUEqh+yc83vOKiCn7T9vcIGewRr2Pv9DBy5bnXsfeH2kJhQZhVKBRqowc9H5p+vHsVD4QLlcz45wgZ7JHKk+1TwvDYmxf9InZNsTHMbGq9eqgurK9owD/0wyfp4r2NfXvAnfrBIYRpaT5blvd/fhf79Ozxu4y2+j8DZchAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDAQMBAwEDgGxaeGOhDPTTVAAAAAElFTkSuQmCC" /><!-- --></p>
<p><strong>Figure 1:</strong> A simple dendrogram with three terminal leaf nodes</p>
<p>As demonstrated in this example, manually setting attributes on dendrogram objects can be rather tedious, motivating the development of functions to automate the generation and manipulation of these tree structures.</p>
</div>
</div>
<div id="the-phylogram-package" class="section level2">
<h2>The ‘phylogram’ package</h2>
<p>Here, we introduce <strong>phylogram</strong>, an R package for working with evolutionary trees as deeply-nested lists. The package contains functions for importing and exporting dendrogram objects to and from parenthetic text, and assembling, visualizing, manipulating and rendering trees for publication. These functions are detailed below with examples of their utility.</p>
<div id="importing-and-exporting-trees" class="section level3">
<h3>Importing and exporting trees</h3>
<p>The Newick (a.k.a. New Hampshire) parenthetic text format <span class="citation">(Felsenstein <em>et al.</em>, 1986)</span> is a universal phylogenetic tree representation that is compatible with most tree-editing software. The <strong>phylogram</strong> package features the text parser <code>read.dendrogram</code> that reads a character string or text file in the Newick format and creates a dendrogram object. This function supports weighted edges, labels with special meta-characters (enclosed in single quotation marks), comments (enclosed in square brackets; ignored by the parser), multifuricating nodes, and both rooted and unrooted trees. Inner-node labels are currently ignored; however, the inclusion of “label” attributes for non-leaf nodes will be available in a future version. Objects of class “dendrogram” can be exported as Newick-style parenthetic text using the function <code>write.dendrogram</code>.</p>
<div id="example-2-import-and-export-a-tree-from-a-newick-string" class="section level4">
<h4>Example 2: Import and export a tree from a Newick string</h4>
<p>The simple Newick string in Example 1 can be imported as a dendrogram object using the <code>read.dendrogram</code> function as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(phylogram)
newick &lt;-<span class="st"> &quot;(A,(B,C));&quot;</span>
x &lt;-<span class="st"> </span><span class="kw">read.dendrogram</span>(<span class="dt">text =</span> newick)
x
<span class="co">#&gt; 'dendrogram' with 2 branches and 3 members total, at height 2</span></code></pre></div>
<p>The following command writes the object back to the console in Newick format without edge weights:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.dendrogram</span>(x, <span class="dt">edges =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; [1] &quot;(A,(B,C));&quot;</span></code></pre></div>
<p>The syntax is similar when reading and writing text files, except that the <code>text</code> argument is replaced by <code>file</code>, and a valid file path is passed to the function.</p>
<p>If required, the dendrogram can be converted to an object of class “phylo” using the <code>as.phylo.dendrogram</code> method, and converted back to a dendrogram with <code>as.dendrogram.phylo</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">as.phylo</span>(x)
z &lt;-<span class="st"> </span><span class="kw">as.dendrogram</span>(y)
<span class="kw">identical</span>(x, z)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="tree-editingmanipulation" class="section level3">
<h3>Tree editing/manipulation</h3>
<p>The <strong>phylogram</strong> package features several additional functions to facilitate some of the more common manipulation operations. Leaf nodes and internal branching nodes can be removed using the function <code>prune</code>, which identifies and recursively deletes nodes based on regular expression pattern matching of node “label” attributes. To aid visualization, the function <code>ladder</code> rearranges the tree, sorting nodes by the number of members (analogous to the <code>ladderize</code> function in the <strong>ape</strong> package). Another function aiding in tree visualization is <code>ultrametricize</code>, which resets the “height” attributes of all terminal leaf nodes to zero (note that unlike <code>ape::chronos()</code> there is no mathematical basis for this operation; rather it is merely a visualization aid). The function <code>reposition</code> scales the height of all nodes in a tree by a given constant (passed <em>via</em> the argument <code>shift</code>), and features the option to reset all node heights so that height of the farthest terminal leaf node from the root is zero (by specifying <code>shift = &quot;reset&quot;</code>). The function <code>remidpoint</code> recursively corrects all “midpoint”, “members” and “leaf” attributes following manual editing of a tree.</p>
</div>
<div id="tree-vizualization" class="section level3">
<h3>Tree vizualization</h3>
<p>Publication-quality trees can be generated from dendrogram objects using the <strong>stats</strong> plotting function <code>plot.dendrogram</code>, and the extensive plotting functions available in dendrogram-enhancing packages such as <strong>circlize</strong> <span class="citation">(Gu <em>et al.</em>, 2014)</span> and <strong>dendextend</strong> <span class="citation">(Galili, 2015)</span>. The latter also offers the facility to convert dendrograms to “ggdend” objects, for which many powerful ‘grammar of graphics’ plotting functions are available in the <strong>ggplot2</strong> <span class="citation">(Wickham, 2009)</span> and <strong>ggdendro</strong> <span class="citation">(DeVries and Ripley, 2016)</span> packages. Moreover, there are several advanced plotting options for “phylo” objects in the <strong>ape</strong> package <span class="citation">(Paradis <em>et al.</em>, 2004)</span>, as well as the Bioconductor package <strong>ggtree</strong> <span class="citation">(Yu <em>et al.</em>, 2017)</span>. Given the extensive tree visualization options already available, we elected not to include any additional plotting functions in the <strong>phylogram</strong> package.</p>
</div>
<div id="getting-help" class="section level3">
<h3>Getting help</h3>
<p>If you experience a problem using this package please either raise it as an issue on <a href="http://github.com/shaunpwilkinson/phylogram/issues">GitHub</a>, post it on the <strong>phylogram</strong> <a href="https://groups.google.com/group/phylogram">google group</a> or email the author directly.</p>
</div>
</div>
<div id="acknowledgements" class="section level2">
<h2>Acknowledgements</h2>
<p>This software was developed with funding from a Rutherford Foundation Postdoctoral Research Fellowship from the Royal Society of New Zealand.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-deVries2016">
<p>DeVries,A. and Ripley,B.D. (2016) ggdendro: create dendrograms and tree diagrams using ’ggplot2’.</p>
</div>
<div id="ref-Felsenstein1986">
<p>Felsenstein,J. <em>et al.</em> (1986) The newick tree format, 1986.</p>
</div>
<div id="ref-Galili2015">
<p>Galili,T. (2015) dendextend: An R package for visualizing, adjusting and comparing trees of hierarchical clustering. <em>Bioinformatics</em>, <strong>31</strong>, 3718–3720.</p>
</div>
<div id="ref-Gu2014">
<p>Gu,Z. <em>et al.</em> (2014) circlize implements and enhances circular visualization in R. <em>Bioinformatics</em>, <strong>30</strong>, 2811–2812.</p>
</div>
<div id="ref-Paradis2004">
<p>Paradis,E. <em>et al.</em> (2004) APE: analyses of phylogenetics and evolution in R language. <em>Bioinformatics</em>, <strong>20</strong>, 289–290.</p>
</div>
<div id="ref-RCoreTeam2015">
<p>R Core Team (2017) R: A Language and Environment for Statistical Computing.</p>
</div>
<div id="ref-Revell2012">
<p>Revell,L.J. (2012) phytools: an R package for phylogenetic comparative biology (and other things). <em>Methods in Ecology and Evolution</em>, <strong>3</strong>, 217–223.</p>
</div>
<div id="ref-Schliep2011">
<p>Schliep,K.P. (2011) phangorn: phylogenetic analysis in R. <em>Bioinformatics</em>, <strong>27</strong>, 592–593.</p>
</div>
<div id="ref-Wickam2009">
<p>Wickham,H. (2009) ggplot2: Elegant Graphics for Data Analysis Springer-Verlag, New York.</p>
</div>
<div id="ref-Guangchuang2017">
<p>Yu,G. <em>et al.</em> (2017) ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <strong>8</strong>, 28–36.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
